src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1904,16 +1924,47 @@ def vnc_bind_choices(self):
     Method called when system is ready, supposed to start VMs
     flagged that way.
     """
-    if args['id'] != 'ready':
-        return
+    async def start_vm(mw, vm):
+        await mw.call('vm.start', vm['id'])
+
+    async def stop_vm(mw, vm):
+        stop_job = await mw.call('vm.stop', vm['id'])
+        await stop_job.wait()
+        if stop_job.error:
+            mw.logger.error(f'Stopping VM {vm["name"]} failed: {stop_job.error}')
+
+        # Let's ensure that we force poweroff the VM if it hasn't stopped by it's defined timeout
+        if (await mw.call('vm.status', vm['id']))['state'] == 'RUNNING':
+            mw.logger.debug(
+                f'Powering off VM {vm["name"]} as it failed to stop within {vm["shutdown_timeout"]} seconds.'
+            )
+            try:
+                await mw.call('vm.poweroff', vm['id'])
+            except Exception as e:
+                mw.logger.debug(f'Powering off VM {vm["name"]} failed: {e}', exc_info=True)
 
-    global ZFS_ARC_MAX_INITIAL
-    ZFS_ARC_MAX_INITIAL = sysctl.filter('vfs.zfs.arc.max')[0].value
+    if args['id'] == 'ready':
+        global ZFS_ARC_MAX_INITIAL
+        ZFS_ARC_MAX_INITIAL = sysctl.filter('vfs.zfs.arc.max')[0].value
 
-    await middleware.call('vm.initialize_vms')
+        await middleware.call('vm.initialize_vms')
 
-    for vm in await middleware.call('vm.query', [('autostart', '=', True)]):
-        await middleware.call('vm.start', vm['id'])
+        await asyncio_map(
+            functools.partial(start_vm, middleware),
+            (await middleware.call('vm.query', [('autostart', '=', True)])), 16
+        )
+    elif args['id'] == 'shutdown':
+        async with SHUTDOWN_LOCK:
+            await asyncio_map(
+                functools.partial(stop_vm, middleware),
+                (await middleware.call('vm.query', [('status.state', '=', 'RUNNING')])), 16
+            )
+            middleware.logger.debug('VM(s) stopped successfully')
+            # We do this in vm.terminate as well, reasoning for repeating this here is that we don't want to
```

Comment:
Hm, I don't see you doing it in `terminate`, am I missing something?

Reply:
In `terminate` i close down the libvirt connection i have been maintaining, but however i do not stop libvirtd itself. That is only done during shutdown